### ISSUE TITLE
Fix S3966 FP: disposing element of indexable after declaration pattern

### ISSUE THREAD

**Issue by bottle2** [SonarSource/sonar-dotnet#8946]
### Description

Warning S3966 is reported when trying to call .Dispose() **inside a for loop** on the converted result of a declaration pattern of an element on an indexable object, such as an array, `List<T>`, `IList<T>`, `Span<T>` or any type with an indexer (e.g. `UnsafeList<T>` of Unity.Collections).

### Repro steps

```C#
FileStream[] fss = [];

DisposeAllFP1(fss);
DisposeAllFP2(fss);
DisposeFirst1(fss);
DisposeFirst2(fss);
DisposeAll1(fss);
DisposeAll2(fss);
DisposeAll3(fss);
DisposeAll4(fss);
MaybeDispose1(fss);
MaybeDispose2(fss);

static void DisposeAllFP1<T>(T[] array)
{
    for (int i = 0; i < array.Length; i++)
        if (array[i] is IDisposable d)
        {
            d.Dispose();
        }
}

static void DisposeAllFP2(object[] array)
{
    for (int i = 0; i < array.Length; i++)
        if (array[i] is IDisposable d)
        {
            d.Dispose();
        }
}

static void DisposeFirst1(object[] o)
{
    if (o.Length >= 1 && o[0] is IDisposable d)
    {
        d.Dispose();
    }
}

static void DisposeFirst2<T>(T[] o)
{
    if (o.Length >= 1 && o[0] is IDisposable d)
    {
        d.Dispose();
    }
}

static void MaybeDispose1(object o)
{
    if (o is IDisposable d)
    {
        d.Dispose();
    }
}

static void MaybeDispose2<T>(T o)
{
    if (o is IDisposable d)
    {
        d.Dispose();
    }
}

static void DisposeAll1<T>(T[] array) where T : IDisposable
{
    for (int i = 0; i < array.Length; i++)
        array[i].Dispose();
}

static void DisposeAll2(IDisposable[] array)
{
    for (int i = 0; i < array.Length; i++)
        array[i].Dispose();
}

static void DisposeAll3(object[] array)
{
    for (int i = 0; i < array.Length; i++)
        if (array[i] is IDisposable)
        {
            ((IDisposable)array[i]).Dispose();
        }
}

static void DisposeAll4<T>(T[] array)
{
    for (int i = 0; i < array.Length; i++)
        if (array[i] is IDisposable)
        {
            ((IDisposable)array[i]).Dispose();
        }
}
```

### Expected behavior

No warning is reported.

### Actual behavior

The warning S3966 is reported.

### Known workarounds

Cast to `IDisposable` explicitly instead of using the result of declared pattern, or use `foreach`.

### Related information

* C# plugin
* Microsoft Visual Studio Community 2022 (64-bit) version 17.9.3
* ~MSBuild / dotnet version~ I'm tired
* SonarLint for Visual Studio 2022 version 7.7.0.86423 by SonarSource
* Windows 10 version 22H2 compilation 19045.4170

**Comment by bottle2** [comment]
I rewrote my code, this specific pattern does not occur anymore there. you may mark this issue as low priority.

**Comment by gregory-paidis-sonarsource** [comment]
Hey, I can confirm that this is an FP.
Thanks for raising this, I will add it to our backlog.

**Comment by martin-strecker-sonarsource** [comment]
Tracked by https://sonarsource.atlassian.net/browse/NET-1207 now

**Comment by mary-georgiou-sonarsource** [comment]
Fixed in sonar-dotnet 10.9
